File: src/test/TTRP.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Main.java to edit this template
 */
package test;

import cujae.inf.citi.om.data.CustomerType;
import cujae.inf.citi.om.data.ProblemType;
import cujae.inf.citi.om.factory.interfaces.DistanceType;
import cujae.inf.citi.om.factory.interfaces.HeuristicType;
import cujae.inf.citi.om.generator.controller.StrategyHeuristic;
import cujae.inf.citi.om.generator.solution.Solution;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;

/**
 *
 * @author kmych
 */
public class TTRP {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) throws IOException, IllegalArgumentException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException 
    {
        String pathFiles = "instances-ttrp//TTRP_1.txt";
            //int totalInstances = 5;
            LoadFile loadFile = new LoadFile();

            //for(int i = 1; i < totalInstances; i++)
            //{
                loadFile.loadFile(pathFiles); //i + 1

                ArrayList<Integer> idCustomers = new ArrayList<Integer>();
                ArrayList<Double> axisXCustomers = new ArrayList<Double>();
                ArrayList<Double> axisYCustomers = new ArrayList<Double>();
                ArrayList<Double> requestCustomers = new ArrayList<Double>();
                ArrayList<Integer> typeCustomers = new ArrayList<Integer>();

                ArrayList<Integer> idDepots = new ArrayList<Integer>();
                ArrayList<Double> axisXDepots = new ArrayList<Double>();
                ArrayList<Double> axisYDepots = new ArrayList<Double>();
                ArrayList<Integer> countVehicles = new ArrayList<Integer>();
                ArrayList<Double> capacityVehicles = new ArrayList<Double>();
                ArrayList<Integer> countTrailers = new ArrayList<Integer>();
                ArrayList<Double> capacityTrailers = new ArrayList<Double>();

                ArrayList<ArrayList<Double>> listDistances = new ArrayList<ArrayList<Double>>();

                loadFile.loadCountVehiclesForDepotTTRP(countVehicles);
                loadFile.loadCapacityVehiclesTTRP(capacityVehicles);
                loadFile.loadCountTrailersForDepotTTRP(countTrailers);
                loadFile.loadCapacityTrailersTTRP(capacityTrailers);
                loadFile.loadCustomersTTRP(idCustomers, axisXCustomers, axisYCustomers, requestCustomers, typeCustomers);
                loadFile.loadDepotsTTRP(idDepots, axisXDepots, axisYDepots);

                loadFile.fillListDistances(idCustomers, axisXCustomers, axisYCustomers, idDepots, axisXDepots, axisYDepots, listDistances);

                ArrayList<ArrayList<Integer>> idAssignedCustomers = new ArrayList<ArrayList<Integer>>();
                idAssignedCustomers.add(idCustomers);

                HeuristicType heuristicType = HeuristicType.SaveParallel;
                
                if(heuristicType == HeuristicType.Sweep){
                    if(StrategyHeuristic.getStrategyHeuristic().loadProblem(idCustomers, requestCustomers, axisXCustomers, axisYCustomers,
                            typeCustomers, idDepots, axisXDepots, axisYDepots, idAssignedCustomers, countVehicles, capacityVehicles, 
                            countTrailers, capacityTrailers, ProblemType.TTRP, DistanceType.Euclidean))
                    {
                                StrategyHeuristic.getStrategyHeuristic().executeHeuristic(20, heuristicType);
                                Solution result = StrategyHeuristic.getStrategyHeuristic().getBestSolution();
                                double cost = StrategyHeuristic.getStrategyHeuristic().getTotalCostSolution();
                                int requestByRoute = StrategyHeuristic.getStrategyHeuristic().getRequestByRoute().size();
                                long time = StrategyHeuristic.getStrategyHeuristic().getTimeExecute();

                                System.out.println(" ");
                                System.out.println("------------------------------------------");
                                //System.out.println("INSTANCIA: P" + (i + 1));
                                System.out.println("HEURÍSTICA DE CONSTRUCCIÓN: " + heuristicType);
                                System.out.println("COSTO TOTAL: " + cost);
                                System.out.println("TOTAL DE RUTAS: " + requestByRoute);
                                //System.out.println("TIEMPO DE EJECUCIÓN: " + time);
                                System.out.println(" ");
                                for(int j = 0; j < requestByRoute; j++)
                                    System.out.println("R" + (j+1) + result.getListRoutes().get(j).getListIdCustomers());
                                System.out.println("------------------------------------------");
                    }
                    
                }
                else{
                    if(StrategyHeuristic.getStrategyHeuristic().loadProblem(idCustomers, requestCustomers, typeCustomers, idDepots, idAssignedCustomers, countVehicles, capacityVehicles, countTrailers, capacityTrailers, listDistances, ProblemType.TTRP))
                        {
                                StrategyHeuristic.getStrategyHeuristic().executeHeuristic(20, heuristicType);
                                Solution result = StrategyHeuristic.getStrategyHeuristic().getBestSolution();
                                double cost = StrategyHeuristic.getStrategyHeuristic().getTotalCostSolution();
                                int requestByRoute = StrategyHeuristic.getStrategyHeuristic().getRequestByRoute().size();
                                long time = StrategyHeuristic.getStrategyHeuristic().getTimeExecute();

                                System.out.println(" ");
                                System.out.println("------------------------------------------");
                                //System.out.println("INSTANCIA: P" + (i + 1));
                                System.out.println("HEURÍSTICA DE CONSTRUCCIÓN: " + heuristicType);
                                System.out.println("COSTO TOTAL: " + cost);
                                System.out.println("TOTAL DE RUTAS: " + requestByRoute);
                                //System.out.println("TIEMPO DE EJECUCIÓN: " + time);
                                System.out.println(" ");
                                for(int j = 0; j < requestByRoute; j++)
                                    System.out.println("R" + (j+1) + result.getListRoutes().get(j).getListIdCustomers());
                                System.out.println("------------------------------------------");
                        }
                }
                
    }
    
}

File: _BHCVRP_Python_Version/BHCVRP_Python_Version/cujae_inf_citi_om/TTRP.py
import sys
import os
import sys
from tools.LoadFile import LoadFile
from data.ProblemType import ProblemType
from factory.interfaces.HeuristicType import HeuristicType
from generator.controller.StrategyHeuristic import StrategyHeuristic
from factory.interfaces.DistanceType import DistanceType

def main():
    try:
        file_output_stream = open("../Resultado_TTRP1.txt", "w")
        #sys.stdout = file_output

        path_files = "../TTRP_1.txt"
        load_file = LoadFile()

        load_file.load_file(path_files)

        id_customers = []
        axis_x_customers = []
        axis_y_customers = []
        request_customers = []
        type_customers = []

        id_depots = []
        axis_x_depots = []
        axis_y_depots = []
        count_vehicles = []
        capacity_vehicles = []
        count_trailers = []
        capacity_trailers = []

        list_distances = []

        load_file.load_count_vehicles_for_depot_ttrp(count_vehicles)
        load_file.load_capacity_vehicles_ttrp(capacity_vehicles)
        load_file.load_count_trailers_for_depot_ttrp(count_trailers)
        load_file.load_capacity_trailers_ttrp(capacity_trailers)
        load_file.is_load_customers_ttrp(id_customers, axis_x_customers, axis_y_customers, request_customers, type_customers)
        load_file.is_load_depots_ttrp(id_depots, axis_x_depots, axis_y_depots)

        load_file.fill_list_distances(id_customers, axis_x_customers, axis_y_customers, id_depots, axis_x_depots, axis_y_depots, list_distances)

        id_assigned_customers = []
        id_assigned_customers.extend(id_customers)

        heuristic_type = HeuristicType.MoleJameson
        count_execution = 100

        if heuristic_type == HeuristicType.Sweep:
            if StrategyHeuristic.get_strategy_heuristic().load_problem_(id_customers, request_customers, axis_x_customers, axis_y_customers,
                    type_customers, id_depots, axis_x_depots, axis_y_depots, id_assigned_customers, count_vehicles, capacity_vehicles, 
                    count_trailers, capacity_trailers, ProblemType.TTRP, DistanceType.Euclidean):
                
                StrategyHeuristic.get_strategy_heuristic().execute_heuristic(20, heuristic_type)
                result = StrategyHeuristic.get_strategy_heuristic().get_best_solution()
                cost = StrategyHeuristic.get_strategy_heuristic().get_total_cost_solution()
                request_by_route = StrategyHeuristic.get_strategy_heuristic().get_request_by_route().size()
                time = StrategyHeuristic.get_strategy_heuristic().get_time_execute()

                print(" ")
                print("------------------------------------------")
                print("HEURÍSTICA DE CONSTRUCCIÓN: " + heuristic_type)
                print("COSTO TOTAL: " + cost)
                print("TOTAL DE RUTAS: " + request_by_route)
                print(" ")

                for j in range(request_by_route):
                    print("R" + str(j+1) + result.get_list_routes()[j].get_list_id_customers())
                
                print("------------------------------------------")
                
        elif StrategyHeuristic.get_strategy_heuristic().load_problem(id_customers, request_customers, type_customers, id_depots, id_assigned_customers, count_vehicles, capacity_vehicles, count_trailers, capacity_trailers, list_distances, ProblemType.TTRP):
            StrategyHeuristic.get_strategy_heuristic().execute_heuristic(20, heuristic_type)
            result = StrategyHeuristic.get_strategy_heuristic().get_best_solution()
            cost = StrategyHeuristic.get_strategy_heuristic().get_total_cost_solution()
            request_by_route = len(StrategyHeuristic.get_strategy_heuristic().get_request_by_route())
            time = StrategyHeuristic.get_strategy_heuristic().get_time_execute()

            print(" ")
            print("------------------------------------------")
            print("HEURÍSTICA DE CONSTRUCCIÓN: " + heuristic_type.name)
            print("COSTO TOTAL: " + str(cost))
            print("TOTAL DE RUTAS: " + str(request_by_route))
            print(" ")

            for j in range(request_by_route):
                print("R" + str(j + 1) + str(result.get_list_routes()[j].get_list_id_customers()))
                #print("R" + str(j+1) + result.get_list_routes()[j].get_list_id_customers())
                
            print("------------------------------------------")

        file_output_stream.close()
        #sys.stdout = sys.__stdout__

    except IOError as e:
        print(e)

if __name__ == "__main__":
    main()